src/ppx_variants_conv.ml

```diff
@@ -4,619 +4,830 @@
-          |> List.unzip
-        in
-        ( ppat_construct
-            ~loc
-            (Located.lident ~loc name)
-            (Some (ppat_record ~loc patterns Closed))
-        , idents )
-    in
-    let ident_expr = function
-      | `Unlabelled ident -> evar ~loc ident
-      | `Labelled (label, ident) -> pexp_variant ~loc label (Some (evar ~loc ident))
-    in
-    let expr =
-      match idents with
-      | [] -> [%expr ()]
-      | idents -> pexp_tuple ~loc (List.map idents ~f:ident_expr)
-    in
-    pat, expr
-end
-
-let variant_name_to_string v =
-  let s = String.lowercase v in
-  if Keyword.is_keyword s
-  then s ^ "_"
-  else s
-
-module Inspect = struct
-  let row_field loc rf : Variant_constructor.t =
-    match rf.prf_desc with
-    | Rtag ({ txt = name; _ }, true, _) | Rtag ({ txt = name; _ }, _, []) ->
-      { name
-      ; loc
-      ; kind = `Polymorphic None
-      }
-    | Rtag ({ txt = name; _}, false, tp :: _) ->
-      { name
-      ; loc
-      ; kind = `Polymorphic (Some tp)
-      }
-    | Rinherit _ ->
-      Location.raise_errorf ~loc
-        "ppx_variants_conv: polymorphic variant inclusion is not supported"
-
-  let constructor cd : Variant_constructor.t =
-    if Option.is_some cd.pcd_res then
-      Location.raise_errorf ~loc:cd.pcd_loc "GADTs are not supported by variantslib";
-    let kind =
-      match cd.pcd_args with
-      | Pcstr_tuple pcd_args -> `Normal pcd_args
-      | Pcstr_record fields  -> `Normal_inline_record fields
-    in
-    { name = cd.pcd_name.txt
-    ; loc = cd.pcd_name.loc
-    ; kind
-    }
-
-  let type_decl td =
-    let loc = td.ptype_loc in
-    match td.ptype_kind with
-    | Ptype_variant cds ->
-      let cds = List.map cds ~f:constructor in
-      let names_as_string = Hashtbl.create (module String) in
-      List.iter cds ~f:(fun { name; loc; _ } ->
-        let s = variant_name_to_string name in
-        match Hashtbl.find names_as_string s with
-        | None -> Hashtbl.add_exn names_as_string ~key:s ~data:name
-        | Some name' ->
+   open Base
+   open Ppxlib
+   open Ast_builder.Default
+   
+   let raise_unsupported loc =
+     Location.raise_errorf ~loc
+       "Unsupported use of variants (you can only use it on variant types)."
+
+   let raise_unsupported_ty loc =
+     Location.raise_errorf ~loc
+       "Use of this type is not supported use by variants."
+   
+   module Create = struct
+     let lambda loc xs body =
+       List.fold_right xs ~init:body ~f:(fun (label, p) e -> pexp_fun ~loc label None p e)
+     ;;
+   
+     let lambda_sig loc arg_tys body_ty =
+       List.fold_right arg_tys ~init:body_ty ~f:(fun (label, arg_ty) acc ->
+         ptyp_arrow ~loc label arg_ty acc)
+     ;;
+   end
+
+   module Core_type = struct
+
+     let tyvar_name i =
+       let c, n = (Int.rem i 26, i / 26) in
+       let a = String.of_char @@ Char.of_int_exn (c + 97)
+       and v = if n = 0 then "" else Int.to_string n in
+       a ^ v
+     ;;
+
```

Comment:
Pretty sure ppxlib already provides this as gen_symbol.

Reply:
Today I learned it does!